src/Plutarch/Extra/StateThread.hs

```diff
@@ -0,0 +1,57 @@
+module Plutarch.Extra.StateThread (
+  withStateThread,
+) where
+
+import Plutarch.Api.V1 (PCurrencySymbol, PValue)
+import Plutarch.Api.V1.Value (pvalueOf)
+import Plutarch.Api.V2 (
+  AmountGuarantees,
+  KeyGuarantees,
+  PScriptContext,
+  PScriptPurpose (PMinting),
+  PTxInInfo,
+ )
+import Plutarch.Extra.Field (pletAll)
+import PlutusLedgerApi.V2 (TxOutRef)
+
+{- | Adds a state thread to a minting policy.
+
+ @since 3.9.2
+-}
+withStateThread ::
+  forall (s :: S).
+  -- | Initial spend
+  TxOutRef ->
+  -- | Minting policy to wrap
+  Term s (PData :--> PScriptContext :--> POpaque) ->
+  Term s (PData :--> PScriptContext :--> POpaque)
+withStateThread ref mp = plam $ \red ctx -> pletAll ctx $ \ctx' ->
+  pletFields @'["inputs", "mint"] (getField @"txInfo" ctx') $ \txInfo ->
+    pmatch (getField @"purpose" ctx') $ \case
+      PMinting thisPolicy ->
+        pif
+          (uniqueStateTokenMint (pfield @"_0" # thisPolicy) . getField @"mint" $ txInfo)
+          ( pif
+              (pany # hasUniqueInput ref # getField @"inputs" txInfo)
+              (mp # red # ctx)
+              (ptraceError "stateThread: Input not unique")
```

Comment:
I would change this error to something more descriptive, like "Unique input not found"